src/backend/distributed/master/master_delete_protocol.c

```diff
@@ -163,7 +167,208 @@ master_apply_delete_command(PG_FUNCTION_ARGS)
+				char *nodeUser = CurrentUserName();
+				PGconn *connection = ConnectToNode(workerName, workerPort, nodeUser);
+				TransactionConnection *transactionConnection = NULL;
+				PGresult *result = NULL;
+				char *placementAffectedTupleString = NULL;
+				int32 placementAffectedTupleCount = -1;
+
+				if (connection == NULL)
+				{
+					shardFailedPlacementList = lappend(shardFailedPlacementList,
+													   shardPlacement);
+					continue;
+				}
+
+				/* send the query */
+				result = PQexec(connection, "BEGIN");
+				if (PQresultStatus(result) != PGRES_COMMAND_OK)
+				{
+					ReportRemoteError(connection, result);
+					shardFailedPlacementList = lappend(shardFailedPlacementList,
+													   shardPlacement);
+					continue;
+				}
+
+				result = PQexec(connection, shardQueryString->data);
+				if (PQresultStatus(result) != PGRES_COMMAND_OK)
+				{
+					ReportRemoteError(connection, result);
+					shardFailedPlacementList = lappend(shardFailedPlacementList,
+													   shardPlacement);
+					continue;
+				}
+
+				placementAffectedTupleString = PQcmdTuples(result);
+				placementAffectedTupleCount = pg_atoi(placementAffectedTupleString,
+													  sizeof(int32), 0);
+
+				if ((shardAffectedTupleCount == -1) ||
+					(shardAffectedTupleCount == placementAffectedTupleCount))
+				{
+					shardAffectedTupleCount = placementAffectedTupleCount;
+				}
+				else
+				{
+					ereport(WARNING, (errmsg(
+										  "modified %d tuples, but expected to modify %d",
+										  placementAffectedTupleCount,
+										  shardAffectedTupleCount),
+									  errdetail("modified placement on %s:%d",
+												workerName, workerPort)));
+
+					if (placementAffectedTupleCount > shardAffectedTupleCount)
+					{
+						shardAffectedTupleCount = placementAffectedTupleCount;
+					}
+				}
+
+				PQclear(result);
+
+				transactionConnection = palloc0(sizeof(TransactionConnection));
+
+				transactionConnection->connectionId = shardId;
+				transactionConnection->transactionState = TRANSACTION_STATE_OPEN;
+				transactionConnection->connection = connection;
+
+				connectionList = lappend(connectionList, transactionConnection);
+			}
+
+			if (list_length(shardFailedPlacementList) == list_length(shardPlacementList))
+			{
+				ereport(ERROR, (errmsg("Could not delete on any of the placements")));
+			}
+
+			affectedTupleCount += shardAffectedTupleCount;
+			tableFailedPlacementList = list_concat(tableFailedPlacementList,
+												   shardFailedPlacementList);
+		}
+
+		if (MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC)
+		{
+			PrepareRemoteTransactions(connectionList);
+		}
+	}
+	PG_CATCH();
+	{
+		/* roll back all transactions */
+		AbortRemoteTransactions(connectionList);
+		CloseConnections(connectionList);
+
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	CommitRemoteTransactions(connectionList, false);
+
+	/* mark failed placements as inactive */
+	foreach(tableFailedPlacementCell, tableFailedPlacementList)
```

Comment:
I'm leaning towards not having failed placement logic in this function and just rolling back if we fail to connect to a node. This is more of a maintenance function, rather than something that would be called from the application in response to a user event. If it cannot complete because of a failed node, it's unlikely to cause a user-facing availability problem. Part of the reason is that our failed placement logic is a bit dodgy right now and could trigger an outage.